rdm/0007-access-control.md

```diff
@@ -0,0 +1,222 @@
+
+**Generators**
+
+A *generator* implements a rule to allow or deny the access to perform an action.  
+
+The generator would return a set of *needs* and *excludes* for each action. The *needs* are the constraints the user has to meet in order to be allowed to perform an action (e.g. be the owner of the record). The *excludes* specify a set of blacklisting constraints, which have precedence over the *needs*. 
+
+For example, the *record owners* generator would look like:
+
+``` python
+Document:
+    _owners: [1, 2, 3]
+
+RecordOwner:
+    needs:
+        return Document._owners
+    excludes:
+        pass
+```
+
+Since the record owner does not set any specific blacklist (*exclude*), the generator would result in an access granted for users with id 1, 2 or 3.
+
+In addition, generators resturn a set of *query filters*, which are used only when searching. The set of rules returned there are in the form of an Elasticsearch Query object. The `RecordOwner` generator would therefore look like:
+
+``` python
+RecordOwner:
+    needs:
+        return Document._owners
+    excludes:
+        pass
+    query_filter:
+        return Query(owner=user.id)
+```
+
+As a result, it can be seen that the *query filter* inverts the paradigm with respect to the *needs*/*excludes*. Instead of receiving a record and returning its owners, it takes the id of the user performing the action and returns a query object. This is due to the nature of the *list* (search) operation. The check must not be *Can user A access record R, for each one of the records in the system*. This would not be scalable. Therefore, the check must be *which documents can be accessed by the user A*.
+
+Invenio RDM comes with a set of predefined generators:
+- Any user: Any user can perform the action.
+- Any user if public: Any user can perform the action if the record is public.
+- Record owners: Owners of the record can access perform the action.
+- Curators: Curators can perform an action.
+- Super Curators: ??
+
+
+**Permission Policies**
+
+A *permission policy* defines the set of generators upon which a certain action can be performed. A record policy would be set as:
+
+``` python
+RecordPermissionPolicy:
+    can_create: []
+    can_read: []
+    can_update: []
+    can_delete: []
+    can_list: []
+```
+
+The policy would return a set of *needs* and *excludes* for each action. The *needs* and *excludes* of a policy is the union of all those of the specified generators.
+
+For the following user story:
+
+```
+As an admin I want to grant read access to the owners of a document, except if the owner is part of team A.
+```
+
+The following policy applies the `RecordOwner` and `TeamMembership` generators to the record read action. The `RecordOwner` generator, allows the owners of the document (defined right above it, with ids 1, 2 and 3) to access the document. On the other hand, `TeamMembershipt` blacklists users belonging to Team A of accesing the document.
```

Comment:
```suggestion
The following policy applies the `RecordOwner` and `NotTeamAMember` generators to the record read action. The `RecordOwner` generator, allows the owners of the document (defined right above it, with ids 1, 2 and 3) to access the document. On the other hand, `NotTeamAMember` blacklists users belonging to Team A.
```

Reply:
Hmm... I still like the `TeamMembership`. It is more generic and shows that it could also be used to explicitly accept `TeamB`, `TeamC`. Setting the name as `NotTeamAMember` would make it quite tied to one team use case and would require 2 more generators for the other cases.

Reply:
Got it. Let's just make sure `TeamMembership` is used throughout (small typos):

```suggestion
The following policy applies the `RecordOwner` and `TeamMembership` generators to the record read action. The `RecordOwner` generator, allows the owners of the document (defined right above it, with ids 1, 2 and 3) to access the document. On the other hand, `TeamMembership` blacklists users belonging to Team A.
```